src/mcts/node.h

```diff
@@ -69,12 +70,14 @@ class Edge {
   // is false) or as opponent (if as_opponent is true).
   Move GetMove(bool as_opponent = false) const;
 
-  // Returns value of Move probability returned from the neural net
-  // (but can be changed by adding Dirichlet noise).
-  float GetP() const { return p_; }
-
-  // Sets move probability.
-  void SetP(float val) { p_ = val; }
+  // Returns or sets value of Move policy prior returned from the neural net
+  // (but can be changed by adding Dirichlet noise). We use a 16 bit significand
+  // for storage savings in memory-critical Edges. Priors are floored at
+  // 1/0xFFFF, since we use priors multiplicatively to calculate U.
+  float GetP() const { return   static_cast<float>(p_)
+                              / static_cast<float>(0xFFFF); }
+  void SetP(float val) { p_ = std::max(static_cast<uint16_t>(1),
```

Comment:
That way p_ can never be 0, kind of a waste of capacity.

What we would want is rounding up:
(example with if we had only 2 bits to store rather than 16)
```
0       0.25       0.5       0.75       1
|---------|---------|---------|---------|
A    B    C     D   E   F   G   H   I   J
```
ABC would be encoded as 0 (0.25 when converted back)
ED as 1  (0.5)
FG as 2 (0.75)
HIJ as 3 (1.0)

For that:
GetP should return ` (p_ + 1) / static_cast<float>(0x4)` (or 0x10000 in case of 16 bit).

SetP should round up. It's more compicated, but it's called much rarer, so probably something like:
```cpp
  auto x = ceil(val * 0x10000);
  p_ = x ? (x-1) : 0;
```

Reply:
Hm... Or probably it's better those 4 bits to mean:
0.125,  0.375, 0.625, 0.875..
and round do the closest.
That way overall error will be less. (and also expected sum of probabilities will remain 1.0 which is good)

Reply:
I thought the conclusion on Discord was that it should not be possible for policy to be 0, because that would mess up the computation of `U` and thus significantly alter `Search` behavior.

Reply:
Yes, with rounding up it's never 0.

Reply:
Ah, the other reason I had a value off limits for policy: how do we distinguish uninitialized-policy Edges from initialized-policy ones? Though I guess the code had never cared about that before (except for one case in PrefetchIntoCache), everywhere just assumes that policy is non-zero, perhaps we should change that?

Reply:
There is one thing to keep in mind here. Ideally, original order should be preserved. So if for example two moves are rounded to the same policy value., that would be suboptimal. So if a > b before rounding it should also be a > b after rounding. There is some Literature on this, and some even claim that order is more important than actual value. I would not go that far, but its something to keep in mind. There are algos which ensure that rounding does not change (strict) order, but i think thats would be clutter the codebase even more. 

Also optimization vs. code clarity is an issue, and instead of casting here and there and working with a number of different datatypes, I would suggest to for example just use FP16 throughout the code.

Reply:
Another idea, given that smaller values much more frequent and need better precision:

```cpp
void SetP(float p) { p = static_cast<uint16_t>(reinterpret_cast<uint32_t>(p) >> 13); }
float GetP() { return reinterpret_cast<float>(static_cast<uint32_t>(p_ << 13)); }
```

The whole bit is wasted to denote value 1.0f, but at least conversion is fast and resulting possible value distribution is really good!

On this site:
https://www.h-schmidt.net/FloatConverter/IEEE754.html

Those are those bits:
![image](https://user-images.githubusercontent.com/11266455/43021335-9ac32106-8c63-11e8-8c6b-f7f6f1099611.png)
Green arrow is a wasted bit, only set for values between 1.00 and 1.999999999999.
We can throw it away and replace with one more bit of mantissa (by more bit magic or by multiplying/dividing value by 2).

Reply:
Instead of chopping the low bits, you can add 0x2000 to sort-off round to nearest. 
```
void SetP(float p) { p = static_cast<uint16_t>((reinterpret_cast<uint32_t>(p) + 0x2000) >> 13); }
float GetP() { return reinterpret_cast<float>(static_cast<uint32_t>(p_ << 13)); }
```

Reply:
It's a good idea with adding 0x2000, but I think it's better to do that in GetP() rather than in SetP().
```cpp
void SetP(float p) { p = static_cast<uint16_t>(reinterpret_cast<uint32_t>(p) >> 13); }
float GetP() { return reinterpret_cast<float>(static_cast<uint32_t>((p_ << 13) + 0x2000)); }
```

Reply:
I think doing it in SetP is more accurate because it is a very good approximation of round to nearest. The only difference is when the 14 lowest bits are exactly equal to 0x2000, where real round to nearest will round up or down depending on the next higher bit.

Reply:
Suppose we have two bit mantissa, then those numbers:
```
0       0.25       0.5       0.75       1  (× 2ⁿ)
|---------|---------|---------|---------|
   A   B     C    D    E   F    G     H
```
With adding 0x2000 in GetP() will become
```
0       0.25       0.5       0.75       1  (× 2ⁿ)
|---------|---------|---------|---------|
     A         C         E         G
     B         D         F         H
```
And with adding 0x2000 in SetP() they will be:
```
0       0.25       0.5       0.75       1  (× 2ⁿ)
|---------|---------|---------|---------|
A         C         E         G
          B         D         F         H
```
Hm... Do you think that the trick with overflowing mantissa into the exponent (H) will always work?

Reply:
Yes, overflowing the mantissa in the exponent works in ieee standard representation, with the exception of inf which is never the case here. I always assumed this is by design to simplify rounding hardware.